### ISSUE TITLE
FEATURE: push commands

### ISSUE THREAD

**Issue by ma-anwar** [talonhub/community#140]
Commands to go to the end of the line and insert some sort of symbol, relative to the language being coded in may be quite useful.
For example for python, the following command may be quite useful,
```
push comma:
     edit.line_end()
     insert(",")
     key(enter)
```
This sort of command can be especially useful for languages like C, where brackets need to be commonly inserted at the end of lines and where semicolons are inserted at the end of statements.

Just wondering if this seems like a good idea or better for users to implement for their specific use cases.

**Comment by auphofBSF** [comment]
I was recently shown this as a generic handler https://github.com/knausj85/knausj_talon/blob/d8736b93413ca38a37ee2865117bb73f31253956/lang/talon.talon#L24, 

I dont know what the collection for punctution is but I think it is defined here https://github.com/knausj85/knausj_talon/blob/d8736b93413ca38a37ee2865117bb73f31253956/code/keys.py#L15 and so I believe could be written (I did not test)

```
push <user.symbol>:
     edit.line_end()
     insert("{symbol}")
     key(enter)
```

**Comment by fidgetingbits** [comment]
Sounds good to me. I use something like this in my vim files to start editing at the end of a line, albeit to me `push` always just implied adding something and/or editing at the end of the line (like append) and doesn't imply adding an enter press to go to the next line. I can just switch my habit to actually use `append` for what I currently use `push` for though, and what you propose will fit in nicely.

I agree just having a generic `push` command as @auphofBSF showed above that adds any symbol would be best.

**Comment by knausj85** [comment]
Yes, this seems like a good idea. @auphofBSF‘s generic command is my preference

**Comment by ma-anwar** [comment]
Sounds good, the generic version works great. Perhaps @auphofBSF  can PR although not sure where it would be appropriate to add. Programming.talon?

**Comment by ma-anwar** [comment]
Was thinking about this further, @fidgetingbits would you have an alternative suggestion for push in terms of what it should be called?

**Comment by fidgetingbits** [comment]
i don't have any other suggestions, i am pretty sure other people use `push` for the same thing, maybe with the exception of adding a new line after. i'm pretty right i picked it up from aenea originally

**Comment by barrcodes** [comment]
I was thinking similar things for inline arrays. Maybe I missed it, but is there a command for `, `

**Comment by ym-han** [comment]
I map 'comspace' to ``comma, and then space`` in my personal talon script

**Comment by knausj85** [comment]
I use spamma
https://github.com/knausj85/knausj_talon/blob/master/text/symbols.talon#L10

**Comment by knausj85** [comment]
I was thinking it would be nice to have something similar for `word` that adds a space after the word. maybe `wordy`

**Comment by barrcodes** [comment]
I can see that. Sometimes I talk too slow during dictation, and then I end up in this weird process of `say word` `space` `say another word`

**Comment by ma-anwar** [comment]
So for words with spaces after, I end up using the phrase command for phrases as usual and I remapped "say" to a command that adds spaces after. It's just a slightly modified version of what phrase was mapped to,
```
def spaceText(m) -> str:
    words = str(m).split(' ')
    
    i = 0
    while i < len(words):
        words[i] = remove_dragon_junk(words[i])
        i += 1

    return (' '.join(words) + ' ')
```
For commas with a space I was using "kick", I never noticed spamma though, I'll start using that.

**Comment by fidgetingbits** [comment]
i also use `spamma` for `, `. 

since i find myself using `push` a lot for appending to lines still the last couple days and thinking about it `push` sort of implies your keep adding things to the line? I wonder if we can instead consider a command that will add a character and then press `enter` as something else like perhaps `final`, `finalize`, etc would be more suiting, then we have `push` work like it already does in my brain (and in other speech repos like aenea iirc).

So I guess atm in my brain:
`append`: start inserting/editing after the current cursor (a vim thing, since append puts you into insert mode. in many editors the append part may be a nop)
`push`: start inserting/editing at the end of the line. so: `push <thing>` == `go to end of line`, `append <thing>`
`final`: start editing at the next line, but after adding something to the end of the line. so: `final <thing>` == `push <thing>`, `key(enter)`

**Comment by ma-anwar** [comment]
Yeah that makes sense, I've ended up using push for inserting at the end of the line as well. That's why I asked again, it makes more sense too. Final seems like a good name for the push + newline, I don't think that will conflict with anything else in the repo iirc as well.

**Comment by knausj85** [comment]
Taking a look at this one now, I'm kinda thinking we might be better off with a short go-to-line-end command vs the default `go line end` in generic_editor.talon.

e.g., maybe add these to [line_commands.talon] (https://github.com/knausj85/knausj_talon/blob/master/text/line_commands.talon) or something. Note line_commands is fairly new, you folks may not have it yet.

```
lend: edit.line_end()
bend: edit.line_start()
```

then we can potentially chain other useful commands stuff like...
`lend comment this is a test`
`lend at sign`
`lend at sign slap` -> finalize equivalent.

to insert a comment at the end of a line, for instance?

**Comment by fidgetingbits** [comment]
i'm good with something like that whether or not we add something like `final` in the end

**Comment by knausj85** [comment]
Adding lend/bend since it seems to work well for me, and keeps the grammar small. 

https://github.com/knausj85/knausj_talon/commit/eeea6454aa217aa5f24d0eb36885c639a439d87c

Leaving this open for feedback

Kinda my thinking:
line_commands.talon : commands for ides/editors, opt-in via tag, may require some code support. Basically modeled after the jetbrains stuff
generic_editor.talon   : global commands for all text boxes

There's some overlap for now I'm not entirely happy with, so things like the below might be moved into generic_editor. Still a WIP. 

```
<user.selection_verbs> way left: user.select_way_left(selection_verbs)
<user.selection_verbs> way right: user.select_way_right(selection_verbs)
```

... but by using a capture, we can add context-specific line commands where appropriate. But it's pretty nice to be using the same commands across three apps (vscode, jetbrains, notepad++).

**Comment by fidgetingbits** [comment]
one thing i guess about a command like `push`/`final` still being useful in addition to `lend` in general is that `push` itself implies you're going to do an insert, so when you're working in something like vim you'd potentially have to say: `lend (insert mode|sit) at sign slap`, whereas `lend` itself just means go to the end of the line.

**Comment by knausj85** [comment]
Oh, vim. Could potentially use a custom action here, defaulted to line start/end for most editors, that vim et al could override to enter insertion mode as well...

**Comment by pokey** [comment]
Going back through issues, and looks like this one dropped off the radar.  I see also the discussion got fairly off topic.

- Is it worth capturing some of the off-topic discussion into one or more other issues?
- Shall we file a PR for the actual requested feature

Fwiw, I have a version of this code, using the term "patch", eg "patch colon":  https://github.com/pokey/pokey_talon/blob/c4bfc67e2920aa4f6093335942d90fa2392bd6e4/text/symbols.talon#L5-L10

**Comment by nriley** [comment]
#2033 appears to more or less address this. Since this issue has sat for over 3 years, please feel free to file a separate issue/PR if you have concerns that #2033 does not address.
